File: DALNEW/Query/CompanyQuery.vb
Imports System.Linq.Expressions
Imports BO
Public Class CompanyQuery

    Public Shared Function GetFilterQeury(filter As CompanyFilter) As Expression(Of Func(Of CompanyInfo, Boolean))
        Dim query As Expression(Of Func(Of CompanyInfo, Boolean)) = GetNameQuery(filter.CompanyName)
        query = query.And(GetActivitesAndQuery(filter.SelectedActivities))
        query = query.And(GetProvincesOrQuery(filter.SelectedProvince))
        Return query
    End Function

    Public Shared Function GetNameQuery(name As String) As Expression(Of Func(Of CompanyInfo, Boolean))
        If String.IsNullOrWhiteSpace(name) Then Return Nothing
        Return Function(w) w.BedrijfsNaam.Contains(name)
    End Function
    Public Shared Function GetActivitesAndQuery(activities As List(Of Integer)) As Expression(Of Func(Of CompanyInfo, Boolean))
        If activities Is Nothing OrElse activities.Count = 0 Then Return Nothing
        Dim query As Expression(Of Func(Of CompanyInfo, Boolean)) = Nothing
        For Each Activity In activities
            Dim id As Integer = Activity
            query = query.And(Function(w) w.Activity.Any(Function(a) a.ActivityID = id))
        Next
        Return query
    End Function
    Public Shared Function GetProvincesOrQuery(provinces As List(Of Integer)) As Expression(Of Func(Of CompanyInfo, Boolean))
        If provinces Is Nothing OrElse provinces.Count = 0 Then Return Nothing
        Return Function(f) f.PostalCode.ProvincieID.HasValue And provinces.Contains(f.PostalCode.ProvincieID)
    End Function

End Class

File: DALCore/Query/CompanyQuery.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Linq.Expressions;
using System.Text;
using System.Threading.Tasks;
using DALCore.Models;
using BOCore;

namespace DALCore.Query
{
    public class CompanyQuery
    {
        public static Expression<Func<CompanyInfo, bool>> GetFilterQeury(CompanyFilter filter)
        {
            Expression<Func<CompanyInfo, bool>> query = GetNameQuery(filter.CompanyName);
            query = query.And(GetActivitesAndQuery(filter.SelectedActivities));
            query = query.And(GetProvincesOrQuery(filter.SelectedProvince));
            return query;
        }

        public static Expression<Func<CompanyInfo, bool>> GetNameQuery(string name)
        {
            if (string.IsNullOrWhiteSpace(name))
                return null/* TODO Change to default(_) if this is not a reference type */;
            return w => w.BedrijfsNaam.Contains(name);
        }
        public static Expression<Func<CompanyInfo, bool>> GetActivitesAndQuery(List<int> activities)
        {
            if (activities == null || activities.Count == 0)
                return null/* TODO Change to default(_) if this is not a reference type */;
            Expression<Func<CompanyInfo, bool>> query = null/* TODO Change to default(_) if this is not a reference type */;
            foreach (var Activity in activities)
            {
                int id = Activity;
                query = query.And(w => w.Activity.Any(a => a.ActivityId == id));
            }
            return query;
        }
        public static Expression<Func<CompanyInfo, bool>> GetProvincesOrQuery(List<int> provinces)
        {
            if (provinces == null || provinces.Count == 0)
                return null/* TODO Change to default(_) if this is not a reference type */;
            return f => f.PostCode.ProvincieId.HasValue & provinces.Contains((int)f.PostCode.ProvincieId);
        }
    }
}
